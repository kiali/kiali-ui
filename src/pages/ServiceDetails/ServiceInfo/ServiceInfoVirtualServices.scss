table.table {
  .status {
    width: 50px;
  }
  .actions {
    width: 120px;
  }
}
