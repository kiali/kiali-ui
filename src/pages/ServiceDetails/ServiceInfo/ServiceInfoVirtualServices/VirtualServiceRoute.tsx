import * as React from 'react';
import * as resolve from 'table-resolver';
import {
  checkForPath,
  Destination,
  DestinationWeight,
  highestSeverity,
  HTTPRoute,
  ObjectCheck,
  ObjectValidation,
  severityToIconName,
  TCPRoute
} from '../../../../types/ServiceInfo';
import { BulletChart, Col, Icon, OverlayTrigger, Popover, Row, Table, Tooltip } from 'patternfly-react';
import DetailObject from '../../../../components/Details/DetailObject';
import { PfColors } from '../../../../components/Pf/PfColors';

interface VirtualServiceRouteProps {
  name: string;
  kind: string;
  routes: any[];
  validations: { [key: string]: ObjectValidation };
}

const PFBlueColors = [
  PfColors.Blue,
  PfColors.Blue500,
  PfColors.Blue600,
  PfColors.Blue300,
  PfColors.Blue200,
  PfColors.Blue100
];

class VirtualServiceRoute extends React.Component<VirtualServiceRouteProps> {
  cellFormat = value => <Table.Cell>{value}</Table.Cell>;
  headerFormat = (label, { column }) => {
    let className = column.property || column.header.label.toLowerCase();
    let colSpan = column.header && column.header.props ? column.header.props.colSpan || '' : '';

    return (
      <Table.Heading colSpan={colSpan} className={className}>
        {label}
      </Table.Heading>
    );
  };

  constructor(props: VirtualServiceRouteProps) {
    super(props);
  }

  columns() {
    return {
      columns: [
        {
          header: {
            label: 'Status',
            formatters: [this.headerFormat],
            props: {
              colSpan: 1
            }
          },
          cell: {
            formatters: [this.cellFormat]
          },
          children: [
            {
              property: 'status.value',
              header: {
                label: '',
                formatters: [this.headerFormat]
              },
              cell: {
                formatters: [this.cellFormat]
              }
            }
          ]
        },
        {
          header: {
            label: 'Destination',
            formatters: [this.headerFormat],
            props: {
              colSpan: 3
            }
          },
          cell: {
            formatters: [this.cellFormat]
          },
          children: [
            {
              property: 'destination.host',
              header: {
                label: 'Host',
                formatters: [this.headerFormat]
              },
              cell: {
                formatters: [this.cellFormat]
              }
            },
            {
              property: 'destination.subset',
              header: {
                label: 'Subset',
                formatters: [this.headerFormat]
              },
              cell: {
                formatters: [this.cellFormat]
              }
            },
            {
              property: 'destination.port',
              header: {
                label: 'Port',
                formatters: [this.headerFormat]
              },
              cell: {
                formatters: [this.cellFormat]
              }
            }
          ]
        },
        {
          header: {
            label: 'Weights',
            formatters: [this.headerFormat],
            props: {
              colSpan: 1
            }
          },
          cell: {
            formatters: [this.cellFormat]
          },
          children: [
            {
              property: 'weight.value',
              header: {
                label: '',
                formatters: [this.headerFormat]
              },
              cell: {
                formatters: [this.cellFormat]
              }
            }
          ]
        }
      ]
    };
  }

  rows(route: any) {
    return (route.route || []).map((routeItem, u) => ({
      id: u,
      status: { value: this.statusFrom(this.validation(), routeItem, u) },
      weight: { value: routeItem.weight ? routeItem.weight : '-' },
      destination: this.destinationFrom(routeItem, u)
    }));
  }

  validation(): ObjectValidation {
    return this.props.validations[this.props.name];
  }

  statusFrom(validation: ObjectValidation, routeItem: DestinationWeight, index: number) {
    let checks = checkForPath(validation, 'spec/route[' + index + ']/weight/' + routeItem.weight);
    checks.push(...checkForPath(validation, 'spec/route[' + index + ']/labels'));
    let severity = highestSeverity(checks);
    let iconName = severity ? severityToIconName(severity) : 'ok';
    if (iconName !== 'ok') {
      return (
        <OverlayTrigger
          placement={'right'}
          overlay={this.infotipContent(checks)}
          trigger={['hover', 'focus']}
          rootClose={false}
        >
          <Icon type="pf" name={iconName} />
        </OverlayTrigger>
      );
    } else {
      return '';
    }
  }

  destinationFrom(destinationWeight: DestinationWeight, i: number) {
    let destination = destinationWeight.destination;
    return {
      host: destination.host || '-',
      subset: destination.subset || '-',
      port: destination.port ? destination.port.number || '-' : '-'
    };
  }

  infotipContent(checks: ObjectCheck[]) {
    return (
      <Popover id={this.props.name + '-weight-tooltip'}>
        {checks.map(check => {
          return this.objectCheckToHtml(check);
        })}
      </Popover>
    );
  }

  objectCheckToHtml(object: ObjectCheck) {
    return (
      <Row>
        <Col xs={1}>
          <Icon type="pf" name={severityToIconName(object.severity)} />
        </Col>
        <Col xs={10} style={{ marginLeft: '-20px' }}>
          {object.message}
        </Col>
      </Row>
    );
  }

  bulletChartValues(routes: TCPRoute | HTTPRoute) {
    return (routes.route || []).map((destinationWeight, u) => ({
      value: routes.route.length === 1 ? 100 : destinationWeight.weight,
      title: `${u}_${destinationWeight.weight}`,
      color: PFBlueColors[u % PFBlueColors.length],
      tooltipFunction: () => {
        const badges = this.renderDestination(destinationWeight.destination);
        return (
          <Tooltip id={`${u}_${destinationWeight.weight}`} key={`${u}_${destinationWeight.weight}`}>
            <div className="label-collection">{badges}</div>
          </Tooltip>
        );
      }
    }));
  }

  renderDestination(destination: Destination) {
    return (
      <ul style={{ listStyleType: 'none', paddingLeft: '15px' }}>
        <li>Host: {destination.host || '-'} </li>
        <li>Subset: {destination.subset || '-'} </li>
        <li>Port: {destination.port ? destination.port.number : '-'} </li>
      </ul>
    );
  }

  renderTable(route: any, i: number) {
    const resolvedColumns = resolve.columnChildren(this.columns());
    const resolvedRows = resolve.resolve({
      columns: resolvedColumns,
      method: resolve.nested
    })(this.rows(route));

    return (
      <div key={'bulletchart-wrapper-' + i} style={{ marginTop: '30px' }}>
        <div>
          <BulletChart
            key={'bullet-chart-' + i}
            label="Weight sum"
            stacked={true}
            thresholdWarning={-1}
            thresholdError={-1}
            values={this.bulletChartValues(route)}
            ranges={[{ value: 100 }]}
          />
        </div>
        <Table.PfProvider columns={resolvedColumns} striped={true} bordered={true} hover={true} dataTable={true}>
          <Table.Header headerRows={resolve.headerRows(this.columns())} />
          <Table.Body rows={resolvedRows} rowKey="id" />
        </Table.PfProvider>
      </div>
    );
  }

  render() {
    return (
      <div>
        {(this.props.routes || []).map((route, i) => (
          <div key={'virtualservice-rule' + i} className="row-cards-pf">
            <Row>
              <Col xs={12} sm={12} md={3} lg={3}>
                <DetailObject name={this.props.kind + ' Route'} detail={route} exclude={['route']} />
              </Col>
              <Col xs={12} sm={12} md={5} lg={5}>
                {this.renderTable(route, i)}
              </Col>
            </Row>
            <hr />
          </div>
        ))}
      </div>
    );
  }
}

export default VirtualServiceRoute;
