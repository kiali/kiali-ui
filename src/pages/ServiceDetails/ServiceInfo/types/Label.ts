export default class Label {
  key: string;
  value: string;

  constructor(key: string, value: string) {
    this.key = key;
    this.value = value;
  }

  isValid() {
    return typeof this.key === 'string' && typeof this.value === 'string';
  }
}
