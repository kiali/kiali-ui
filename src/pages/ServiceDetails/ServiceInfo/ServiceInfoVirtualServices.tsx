import * as React from 'react';
import { Col, Row, Table } from 'patternfly-react';
import * as resolve from 'table-resolver';
import { Link } from 'react-router-dom';
import { ObjectValidation, VirtualService } from '../../../types/IstioObjects';
import './ServiceInfoVirtualServices.css';
import LocalTime from '../../../components/Time/LocalTime';
import { ConfigIndicator } from '../../../components/ConfigValidation/ConfigIndicator';

interface ServiceInfoVirtualServicesProps {
  virtualServices?: VirtualService[];
  validations: { [key: string]: ObjectValidation };
}

class ServiceInfoVirtualServices extends React.Component<ServiceInfoVirtualServicesProps> {
  headerFormat = (label, { column }) => <Table.Heading className={column.property}>{label}</Table.Heading>;
  cellFormat = (value, { column }) => {
    const props = column.cell.props;
    const className = props ? props.align : '';

    return <Table.Cell className={className}>{value}</Table.Cell>;
  };

  columns() {
    return {
      columns: [
        {
          property: 'status',
          header: {
            label: 'Status',
            formatters: [this.headerFormat]
          },
          cell: {
            formatters: [this.cellFormat],
            props: {
              align: 'text-center'
            }
          }
        },
        {
          property: 'name',
          header: {
            label: 'Name',
            formatters: [this.headerFormat]
          },
          cell: {
            formatters: [this.cellFormat]
          }
        },
        {
          property: 'createdAt',
          header: {
            label: 'Created at',
            formatters: [this.headerFormat]
          },
          cell: {
            formatters: [this.cellFormat]
          }
        },
        {
          property: 'resourceVersion',
          header: {
            label: 'Resource version',
            formatters: [this.headerFormat]
          },
          cell: {
            formatters: [this.cellFormat]
          }
        },
        {
          property: 'actions',
          header: {
            label: 'Actions',
            formatters: [this.headerFormat]
          },
          cell: {
            formatters: [this.cellFormat]
          }
        }
      ]
    };
  }

  hasValidations(virtualService: VirtualService): boolean {
    // This is insane, but doing return to the clause inside the if will cause compiler failure
    return !!this.props.validations && !!this.props.validations[virtualService.metadata.name];
  }

  validation(virtualService: VirtualService): ObjectValidation {
    return this.props.validations[virtualService.metadata.name];
  }

  overviewLink(virtualService: VirtualService) {
    return (
      <Link
        to={
          '/namespaces/' +
          virtualService.metadata.namespace +
          '/istio/virtualservices/' +
          virtualService.metadata.name +
          '?list=overview'
        }
      >
        {virtualService.metadata.name}
      </Link>
    );
  }

  yamlLink(virtualService: VirtualService) {
    return (
      <Link
        to={
          '/namespaces/' +
          virtualService.metadata.namespace +
          '/istio/virtualservices/' +
          virtualService.metadata.name +
          '?list=yaml'
        }
      >
        View YAML
      </Link>
    );
  }

  rows() {
    return (this.props.virtualServices || []).map((virtualService, vsIdx) => ({
      id: vsIdx,
      status: (
        <ConfigIndicator
          id={vsIdx + '-config-validation'}
          validations={this.hasValidations(virtualService) ? [this.validation(virtualService)] : []}
        />
      ),
      name: this.overviewLink(virtualService),
      createdAt: <LocalTime time={virtualService.metadata.creationTimestamp || ''} />,
      resourceVersion: virtualService.metadata.resourceVersion,
      actions: this.yamlLink(virtualService)
    }));
  }

  renderTable() {
    return (
      <Table.PfProvider columns={this.columns().columns} striped={true} bordered={true} hover={true} dataTable={true}>
        <Table.Header headerRows={resolve.headerRows(this.columns())} />
        <Table.Body rows={this.rows()} rowKey="id" />
      </Table.PfProvider>
    );
  }

  render() {
    return (
      <Row className="card-pf-body">
        <Col xs={12}>{this.renderTable()}</Col>
      </Row>
    );
  }
}

export default ServiceInfoVirtualServices;
