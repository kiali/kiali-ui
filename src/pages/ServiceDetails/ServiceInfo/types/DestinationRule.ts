import { DestinationRuleSpec, K8sMetadata } from '../../../../types/IstioObjects';
import Subset from './Subset';

export default class DestinationRule {
  metadata: K8sMetadata;
  spec: DestinationRuleSpec;
  unformattedField: string;

  constructor(metadata: K8sMetadata, spec: DestinationRuleSpec) {
    this.metadata = metadata;
    this.spec = spec;
  }

  isValid() {
    return this.hasValidName() && this.hasSpecData() && this.hasValidSubsets() && this.hasValidHost();
  }

  hasValidName() {
    if (typeof this.metadata.name !== 'string') {
      this.unformattedField = 'Name';
      return false;
    }

    return true;
  }

  hasSpecData() {
    return this.spec !== null;
  }

  hasValidSubsets() {
    if (!this.spec.subsets) {
      return true;
    }

    let valid = this.spec.subsets instanceof Array;
    valid =
      valid &&
      this.spec.subsets.every((subset, i, ary) =>
        new Subset(subset.name, subset.labels, subset.trafficPolicy).isValid()
      );

    if (!valid) {
      this.unformattedField = 'Subsets';
    }

    return valid;
  }

  hasValidHost() {
    if (typeof this.spec.host !== 'string') {
      this.unformattedField = 'Host';
      return false;
    }

    return true;
  }

  formatValidation() {
    if (!this.isValid()) {
      return {
        message: 'This destination rule has format problems in field ' + this.unformattedField,
        severity: 'error',
        path: ''
      };
    }

    return null;
  }
}
