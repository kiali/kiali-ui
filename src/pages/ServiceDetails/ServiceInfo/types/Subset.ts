import { TrafficPolicy } from '../../../../types/IstioObjects';
import Label from './Label';

export default class Subset {
  name: string;
  labels: { [key: string]: string };
  trafficPolicy: TrafficPolicy;

  constructor(name: string, labels: { [key: string]: string }, trafficPolicy: TrafficPolicy) {
    this.name = name;
    this.labels = labels;
    this.trafficPolicy = trafficPolicy;
  }

  public isValid() {
    return this.hasValidName() && this.hasValidLabels() && this.hasValidTrafficPolicy();
  }

  hasValidName() {
    return typeof this.name === 'string';
  }

  hasValidLabels() {
    const valid = Object.keys(this.labels).every((k, i) => new Label(k, this.labels[k]).isValid());
    return this.labels instanceof Object && valid;
  }

  hasValidTrafficPolicy() {
    return true;
  }
}
