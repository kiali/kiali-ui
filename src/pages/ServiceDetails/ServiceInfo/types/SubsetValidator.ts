import { Subset } from '../../../../types/IstioObjects';

export default class SubsetValidator {
  subset: Subset;

  constructor(subset: Subset) {
    this.subset = subset;
  }

  public isValid() {
    return this.hasValidName() && this.hasValidLabels();
  }

  hasValidName() {
    return this.hasStringType(this.subset.name);
  }

  hasValidLabels() {
    const valid = Object.keys(this.subset.labels).every((k, _i) => this.hasValidLabel(k, this.subset.labels[k]));
    return this.subset.labels instanceof Object && valid;
  }

  hasValidLabel(name: string, value: string): boolean {
    return this.hasStringType(name) && this.hasStringType(value);
  }

  hasStringType(value: any) {
    return typeof value === 'string';
  }
}
