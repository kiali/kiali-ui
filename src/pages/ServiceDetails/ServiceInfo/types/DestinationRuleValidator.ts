import { DestinationRule } from '../../../../types/IstioObjects';
import SubsetValidator from './SubsetValidator';

export default class DestinationRuleValidator {
  destinationRule: DestinationRule;
  unformattedField: string;

  constructor(destinationRule: DestinationRule) {
    this.destinationRule = destinationRule;
    this.unformattedField = 'none';
  }

  isValid() {
    return this.hasValidName() && this.hasSpecData() && this.hasValidSubsets() && this.hasValidHost();
  }

  hasValidName() {
    if (typeof this.destinationRule.metadata.name !== 'string') {
      this.unformattedField = 'Name';
      return false;
    }

    return true;
  }

  hasSpecData() {
    return this.destinationRule.spec !== null;
  }

  hasValidSubsets() {
    if (!this.destinationRule.spec.subsets) {
      return true;
    }

    let valid = this.destinationRule.spec.subsets instanceof Array;
    valid =
      valid && this.destinationRule.spec.subsets.every((subset, _i, _ary) => new SubsetValidator(subset).isValid());

    if (!valid) {
      this.unformattedField = 'Subsets';
    }

    return valid;
  }

  hasValidHost() {
    if (typeof this.destinationRule.spec.host !== 'string') {
      this.unformattedField = 'Host';
      return false;
    }

    return true;
  }

  formatValidation() {
    if (!this.isValid()) {
      return {
        message: 'This destination rule has format problems in field ' + this.unformattedField,
        severity: 'error',
        path: ''
      };
    }

    return null;
  }
}
