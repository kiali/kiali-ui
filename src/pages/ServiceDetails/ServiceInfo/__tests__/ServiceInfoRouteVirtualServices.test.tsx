import * as React from 'react';
import { shallow } from 'enzyme';
import ServiceInfoVirtualServices from '../ServiceInfoVirtualServices';
import { VirtualService } from '../../../../types/IstioObjects';
import { shallowToJson } from 'enzyme-to-json';

const virtualServices: VirtualService[] = [
  {
    metadata: {
      name: 'reviews-default',
      resourceVersion: '1234',
      creationTimestamp: '2018-03-14T10:17:52Z'
    },
    spec: {
      hosts: ['rewiews'],
      gateways: ['reviews'],
      http: [
        {
          route: [
            {
              destination: {
                subset: 'v1',
                host: 'reviews'
              },
              weight: 55
            },
            {
              destination: {
                subset: 'v3',
                host: 'reviews'
              },
              weight: 55
            }
          ]
        }
      ],
      tcp: [
        {
          match: [],
          route: [
            {
              destination: {
                subset: 'v1',
                host: 'reviews'
              },
              weight: 55
            },
            {
              destination: {
                subset: 'v2',
                host: 'reviews'
              },
              weight: 55
            }
          ]
        }
      ]
    }
  }
];

describe('#ServiceInfoVirtualServices render correctly with data', () => {
  it('should render service virtual services', () => {
    const wrapper = shallow(<ServiceInfoVirtualServices virtualServices={virtualServices} validations={{}} />);
    expect(shallowToJson(wrapper)).toBeDefined();
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });
});
