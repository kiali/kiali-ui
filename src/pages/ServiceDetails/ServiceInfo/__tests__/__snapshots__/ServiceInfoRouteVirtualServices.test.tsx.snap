// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#ServiceInfoVirtualServices render correctly with data should render service virtual services 1`] = `
<Row
  bsClass="row"
  className="card-pf-body"
  componentClass="div"
>
  <Col
    bsClass="col"
    componentClass="div"
    xs={12}
  >
    <TablePfProvider
      bordered={true}
      className=""
      columns={
        Array [
          Object {
            "cell": Object {
              "formatters": Array [
                [Function],
              ],
              "props": Object {
                "align": "text-center",
              },
            },
            "header": Object {
              "formatters": Array [
                [Function],
              ],
              "label": "Status",
            },
            "property": "status",
          },
          Object {
            "cell": Object {
              "formatters": Array [
                [Function],
              ],
            },
            "header": Object {
              "formatters": Array [
                [Function],
              ],
              "label": "Name",
            },
            "property": "name",
          },
          Object {
            "cell": Object {
              "formatters": Array [
                [Function],
              ],
            },
            "header": Object {
              "formatters": Array [
                [Function],
              ],
              "label": "Created at",
            },
            "property": "createdAt",
          },
          Object {
            "cell": Object {
              "formatters": Array [
                [Function],
              ],
            },
            "header": Object {
              "formatters": Array [
                [Function],
              ],
              "label": "Resource version",
            },
            "property": "resourceVersion",
          },
          Object {
            "cell": Object {
              "formatters": Array [
                [Function],
              ],
            },
            "header": Object {
              "formatters": Array [
                [Function],
              ],
              "label": "Actions",
            },
            "property": "actions",
          },
        ]
      }
      components={Object {}}
      condensed={false}
      dataTable={true}
      hover={true}
      inlineEdit={false}
      striped={true}
    >
      <Header
        headerRows={
          Array [
            Array [
              Object {
                "cell": Object {
                  "formatters": Array [
                    [Function],
                  ],
                  "props": Object {
                    "align": "text-center",
                  },
                },
                "header": Object {
                  "formatters": Array [
                    [Function],
                  ],
                  "label": "Status",
                },
                "property": "status",
                "props": Object {
                  "rowSpan": 1,
                },
              },
              Object {
                "cell": Object {
                  "formatters": Array [
                    [Function],
                  ],
                },
                "header": Object {
                  "formatters": Array [
                    [Function],
                  ],
                  "label": "Name",
                },
                "property": "name",
                "props": Object {
                  "rowSpan": 1,
                },
              },
              Object {
                "cell": Object {
                  "formatters": Array [
                    [Function],
                  ],
                },
                "header": Object {
                  "formatters": Array [
                    [Function],
                  ],
                  "label": "Created at",
                },
                "property": "createdAt",
                "props": Object {
                  "rowSpan": 1,
                },
              },
              Object {
                "cell": Object {
                  "formatters": Array [
                    [Function],
                  ],
                },
                "header": Object {
                  "formatters": Array [
                    [Function],
                  ],
                  "label": "Resource version",
                },
                "property": "resourceVersion",
                "props": Object {
                  "rowSpan": 1,
                },
              },
              Object {
                "cell": Object {
                  "formatters": Array [
                    [Function],
                  ],
                },
                "header": Object {
                  "formatters": Array [
                    [Function],
                  ],
                  "label": "Actions",
                },
                "property": "actions",
                "props": Object {
                  "rowSpan": 1,
                },
              },
            ],
          ]
        }
      />
      <Body
        onRow={[Function]}
        rowKey="id"
        rows={
          Array [
            Object {
              "actions": <Link
                to="/namespaces/undefined/istio/virtualservices/reviews-default?list=yaml"
              >
                View YAML
              </Link>,
              "createdAt": <LocalTime
                time="2018-03-14T10:17:52Z"
              />,
              "id": 0,
              "name": <Link
                to="/namespaces/undefined/istio/virtualservices/reviews-default?list=overview"
              >
                reviews-default
              </Link>,
              "resourceVersion": "1234",
              "status": <ConfigIndicator
                id="0-config-validation"
                validations={Array []}
              />,
            },
          ]
        }
      />
    </TablePfProvider>
  </Col>
</Row>
`;
